### PR TITLE
Add warning if REST API is accessible through an insecure connection

### PR DESCRIPTION
**Pull request by thaJeztah**
The remote API allows full privilege escalation and is equivalent to having root access on the host. Because of this, the API should never be accessible through an insecure connection (TCP without TLS, or TCP without TLS  verification).

Although a warning is already logged on startup if the daemon uses an insecure configuration, this warning is not very visible (unless someone decides to read the logs).

This patch attempts to make insecure configuration more visible by sending back warnings through the API (which will be printed when using `docker info`).

**- Description for the changelog**

```Markdown
+ Add warning if REST API is accessible through an insecure connection [moby#37684](https://github.com/moby/moby/pull/37684)
```

**- A picture of a cute animal (not mandatory but encouraged)**

cute-ugly-wombat1

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @vdemeester @cpuguy83 @justincormack 

related to https://github.com/moby/moby/pull/37299 (/cc @cyphar)

**Review by vdemeester — APPROVED** [review]
LGTM :frog:

File: daemon/info.go [hunk 1]
```diff
@@ -171,6 +173,41 @@ func (daemon *Daemon) fillSecurityOptions(v *types.Info, sysInfo *sysinfo.SysInf
 	v.SecurityOptions = securityOptions
 }
 
+func (daemon *Daemon) fillAPIInfo(v *types.Info) {
+	cnf := daemon.configStore
+
+	warn := `
```
Comment (cpuguy83) [inline daemon/info.go]: Nit, this should be a constant.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]
Seems ok, but I don't like that it's parsing hosts for the info endpoint, especially since these don't change while the daemon is running.

Since we already have to parse the hosts can we store this info somewhere?

**Comment by thaJeztah** [comment]
Actually; may not even need the `ParseHost()`, because the startup code parses this, and writes it back to Config; https://github.com/moby/moby/blob/cf72051c3746fa60d7e7995c0342eaca91f07bfd/cmd/dockerd/daemon.go#L593-L595

**Comment by thaJeztah** [comment]
@cpuguy83 updated; PTAL

File: daemon/info.go [hunk 2]
```diff
@@ -171,6 +172,32 @@ func (daemon *Daemon) fillSecurityOptions(v *types.Info, sysInfo *sysinfo.SysInf
...
+	for _, host := range cfg.Hosts {
+		// cnf.Hosts is normalized during startup, so should always have a scheme/proto
+		h := strings.SplitN(host, "://", 2)
+		proto := h[0]
+		addr := h[1]
+		if proto != "tcp" {
+			continue
+		}
+		if !cfg.TLS {
+			v.Warnings = append(v.Warnings, fmt.Sprintf("WARNING: API is accessible on http://%s without TLS protection.%s", addr, warn))
```
Comment (cpuguy83) [inline daemon/info.go]: s/without TLS protection/unencrypted/

File: daemon/info.go [hunk 3]
```diff
@@ -171,6 +172,32 @@ func (daemon *Daemon) fillSecurityOptions(v *types.Info, sysInfo *sysinfo.SysInf
...
+			continue
+		}
+		if !cfg.TLSVerify {
+			v.Warnings = append(v.Warnings, fmt.Sprintf("WARNING: API is accessible on https://%s without TLS verification.%s", addr, warn))
```
Comment (cpuguy83) [inline daemon/info.go]: TLS client verification

**Comment by cyphar** [comment]
LGTM, after @cpuguy83's comments are addressed.

File: daemon/info.go [hunk 4]
```diff
@@ -171,6 +172,32 @@ func (daemon *Daemon) fillSecurityOptions(v *types.Info, sysInfo *sysinfo.SysInf
...
+func (daemon *Daemon) fillAPIInfo(v *types.Info) {
+	const warn string = `
+         Access to the remote API is equivalent to root access on the host. Refer
+         to the 'Docker daemon attack surface' section in the documentation for
+         more information: https://docs.docker.com/engine/security/security/#docker-daemon-attack-surface`
+
+	cfg := daemon.configStore
+	for _, host := range cfg.Hosts {
+		// cnf.Hosts is normalized during startup, so should always have a scheme/proto
+		h := strings.SplitN(host, "://", 2)
```
Comment (cpuguy83) [inline daemon/info.go]: Probably fine for this PR, but it would be nice to eliminate this by being able to do something like `host.Protocol()`

Comment (thaJeztah) [reply daemon/info.go]: Yes; tried to keep this PR minimal, but I'm thinking of exposing some of this information (e.g., the configured hosts) in `docker info` as well. I can do some refactoring as part of that 👍

**Comment by thaJeztah** [comment]
Thanks for reviewing @cyphar 🤗

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by andrewhsu** [comment]
Please ignore powerpc build results. The IBM Power infrastructure is experiencing outage at the moment.

**Review by tiborvass — APPROVED** [review]
LGTM

**Pull request merged**